python/ray/util/sgd/v2/migration/sgd-migration-guide.md

```diff
@@ -0,0 +1,224 @@
+
+final_model = trainer.get_model()
+```
+
+you would do
+```python
+class MyTrainingOperator(TrainingOperator):
+	...
+
+def train_func(config):
+	device = torch.device(f"cuda:{sgd.local_rank()}" if
+                  torch.cuda.is_available() else "cpu")
+	if torch.cuda.is_available():
+		torch.cuda.set_device(device)
+	
+	operator_cls = config["operator_cls"]
+	
+	# Set the args to whatever values you want.
+	training_operator = operator_cls(
+		config=config, 
+		world_rank=sgd.world_rank(), 
+		local_rank=sgd.local_rank(), 
+		is_distributed=True,
+		device=device,
+		use_gpu=True,
+		wrap_ddp=True,
+		add_dist_sampler=True 
+
+	training_operator.setup(config)
+
+	for idx in range(config["num_epochs"]):
+		train_loader = training_operator._get_train_loader()
+		# If using DistributedSampler, set the epoch here.
+		train_loader.set_epoch(idx)
+		training_operator.train_epoch(epoch_idx=idx, iter(train_loader))
+
+		validation_loader = training_operator._get_validation_loader()
+		training_operator.validate(iterator=iter(validation_loader))
+	
+	if sgd.world_rank() == 0:
+		return training_operator._get_original_models()
+	else:
+		return None
+
+trainer = Trainer(backend="torch", num_workers=4, use_gpu=True)
+trainer.start()
+results = trainer.run(train_func, config={"operator_cls": MyTrainingOperator, "num_epochs": 10})
+final_model = results[0]			
+```
+
+#### Alternative 2: Custom Training Function
+If you want to no longer use the `TrainingOperator`, you can define your own custom training function, and use only the parts from `TrainingOperator.train_epoch`, `TrainingOperator.setup`, and `TrainingOperator.validate` that are necessary for your application.
+
+You can see a full example on how to port over regular PyTorch DDP code to Ray SGD here: https://docs.ray.io/en/master/raysgd/v2/user_guide.html#update-training-function
+
+### Tensorflow
+The API for `TFTrainer` uses creator functions instead of a `TrainingOperator` to define the training logic. To port over Ray SGD v1 Tensorflow code to v2 you can do the following:
+```python
+def train_func(config):
+	train_dataset, val_dataset = data_creator(config)
+	strategy = MultiWorkerMirroredStrategy()
+	with strategy.scope():
+		model = model_creator(config)
+	
+	for epoch_idx in range(config["num_epochs"]):
+		model.fit(train_dataset)
+	
+	if sgd.world_rank() == 0:
+		return model
+	else:
+		return None
+
+trainer = Trainer(backend="tensorflow", num_workers=4, config={"num_epochs": 3, ...})
+trainer.start()
+model = trainer.run(train_func)
+```
+
+## Interacting with the `Trainer`
+In Ray SGD v1, the user can iteratively call `trainer.train()` or `trainer.validate()` for each epoch, and can then interact with the trainer to get certain state (model, checkpoints, results, etc.). In Ray SGD v2, this is replaced by a single training function that defines the full training & validation loop for all epochs. There are 3 ways to get state during or after the training execution:
```

Comment:
Overall comment - we should either pick the third person perspective ("the user") or the second person one ("you") and stick with it thought the entire doc.

Reply:
Some of this would be good to move to the previous `Training Function` section:
> In Ray SGD v1, the user can iteratively call `trainer.train()` or `trainer.validate()` for each epoch
> In Ray SGD v2, this is replaced by a single training function that defines the full training & validation loop for all epochs.

Reply:
Good point, fixed! @matthewdeng I added a point about this in the API differences.